hier_config/model.py

```diff
@@ -0,0 +1,111 @@
+from enum import Enum, auto
+
+from pydantic import BaseModel as PydanticBaseModel
+from pydantic import ConfigDict, PositiveInt
+
+
+class BaseModel(PydanticBaseModel):
+    """Pydantic.BaseModel with a safe config applied."""
+
+    model_config = ConfigDict(frozen=True, extra="forbid")
+
+
+class DumpLine(BaseModel):
+    depth: PositiveInt
+    text: str
+    tags: frozenset[str]
+    comments: frozenset[str]
+    new_in_config: bool
+
+
+class MatchRule(BaseModel):
+    equals: str | frozenset[str] | None = None
+    startswith: str | tuple[str, ...] | None = None
+    endswith: str | tuple[str, ...] | None = None
+    contains: str | tuple[str, ...] | None = None
+    re_search: str | None = None
+
+
+class TagRule(BaseModel):
+    match_rules: tuple[MatchRule, ...]
```

Comment:
Both `lineage` and `match_rules` are used to describe a `tuple[MatchRule, ...]`. Should we standardize on `lineage`, `match_rule`, or something else?

Reply:
I ended up standardizing on `match_rules`.